File: exercise/python/day23/tests/test_control_system.py
import sys
import unittest
from io import StringIO

from assertpy import assert_that

from core.control import ControlSystem, SleighAction, SleighEngineStatus


class TestControlSystem(unittest.TestCase):
    def setUp(self):
        self.held_output = StringIO()
        sys.stdout = self.held_output

    def tearDown(self):
        self.held_output.close()
        sys.stdout = sys.__stdout__

    def test_start(self):
        # The system has been started
        control_system = ControlSystem()
        control_system.action = SleighAction.FLYING
        control_system.status = SleighEngineStatus.OFF
        control_system.start_system()
        assert_that(control_system.status).is_equal_to(SleighEngineStatus.ON)
        assert_that(self.held_output.getvalue().strip()).is_equal_to("Starting the sleigh...\nSystem ready.")

    def test_ascend(self):
        control_system = ControlSystem()
        control_system.start_system()
        control_system.ascend()
        assert_that(control_system.action).is_equal_to(SleighAction.FLYING)
        assert_that(self.held_output.getvalue().strip()).is_equal_to(
            "Starting the sleigh...\nSystem ready.\nAscending...")

    def test_descend(self):
        control_system = ControlSystem()
        control_system.start_system()
        control_system.ascend()
        control_system.descend()
        assert_that(control_system.action).is_equal_to(SleighAction.HOVERING)
        assert_that(self.held_output.getvalue().strip()).is_equal_to(
            "Starting the sleigh...\nSystem ready.\nAscending...\nDescending...")

    def test_park(self):
        control_system = ControlSystem()
        control_system.start_system()
        control_system.park()
        assert_that(control_system.action).is_equal_to(SleighAction.PARKED)


if __name__ == "__main__":
    unittest.main()

File: exercise/C#/day23/ControlSystem.Tests/TestControlSystem.cs
using ControlSystem.Core;
using FluentAssertions;

namespace ControlSystem.Tests
{
    public class TestControlSystem : IDisposable
    {
        private readonly StringWriter _output;
        private readonly TextWriter _originalOutput;

        public TestControlSystem()
        {
            _output = new StringWriter();
            _originalOutput = Console.Out;
            Console.SetOut(_output);
        }

        [Fact]
        public void TestStart()
        {
            // The system has been started
            var controlSystem = new Core.System();
            controlSystem.Action = SleighAction.Flying;
            controlSystem.Status = SleighEngineStatus.Off;
            controlSystem.StartSystem();
            controlSystem.Status.Should().Be(SleighEngineStatus.On);
            _output.ToString().Trim().Should().Be("Starting the sleigh...\r\nSystem ready.");
        }

        [Fact]
        public void TestAscend()
        {
            var controlSystem = new Core.System();
            controlSystem.StartSystem();
            controlSystem.Invoking(cs => cs.Ascend()).Should().NotThrow<ReindeersNeedRestException>();
            controlSystem.Action.Should().Be(SleighAction.Flying);
            _output.ToString().Trim().Should().Be("Starting the sleigh...\r\nSystem ready.\r\nAscending...");
        }

        [Fact]
        public void TestDescend()
        {
            var controlSystem = new Core.System();
            controlSystem.StartSystem();
            controlSystem.Ascend();
            controlSystem.Invoking(cs => cs.Descend()).Should().NotThrow<SleighNotStartedException>();
            controlSystem.Action.Should().Be(SleighAction.Hovering);
            _output.ToString().Trim().Should().Be("Starting the sleigh...\r\nSystem ready.\r\nAscending...\r\nDescending...");
        }

        [Fact]
        public void TestPark()
        {
            var controlSystem = new Core.System();
            controlSystem.StartSystem();
            controlSystem.Invoking(cs => cs.Park()).Should().NotThrow<SleighNotStartedException>();
            controlSystem.Action.Should().Be(SleighAction.Parked);
        }

        public void Dispose()
        {
            Console.SetOut(_originalOutput);
            _output.Dispose();
        }
    }
}